fetchai/ledger/crypto/entity.py

```diff
@@ -82,17 +97,119 @@ def signing_key(self):
     def sign(self, message: bytes):
         return self._signing_key.sign(message)
 
-    def dumps(self):
-        return json.dumps(self._to_json_object())
+    def prompt_dump(self, fp):
+        password = input("Please enter password")
+        return self.dump(fp, password)
+
+    def dumps(self, password):
+        return json.dumps(self._to_json_object(password))
 
-    def dump(self, fp):
-        return json.dump(self._to_json_object(), fp)
+    def dump(self, fp, password):
+        return json.dump(self._to_json_object(password), fp)
 
-    def _to_json_object(self):
+    def _to_json_object(self, password):
+        encrypted, key_length, init_vec, salt = _encrypt(password, self.private_key_bytes)
         return {
-            'privateKey': base64.b64encode(self._private_key_bytes).decode()
+            'key_length': key_length,
+            'init_vector': init_vec,
+            'password_salt': salt.hex(),
+            'privateKey': base64.b64encode(encrypted).decode()
         }
 
     @classmethod
-    def _from_json_object(cls, obj):
-        return cls.from_base64(obj['privateKey'])
+    def _from_json_object(cls, obj, password):
+        private_key = _decrypt(password,
+                               bytes.fromhex(obj['password_salt']),
+                               base64.b64decode(obj['privateKey']),
+                               obj['key_length'],
+                               obj['init_vector'])
+
+        return cls.from_base64(base64.b64encode(private_key).decode())
+
+
+def _encrypt(password: str, data: bytes) -> Tuple[bytes, int, str, bytes]:
+    """
+    Encryption schema for private keys
+    :param password: plaintext password to use for encryption
+    :param data: plaintext data to encrypt
+    :return: encrypted data, length of original data, initialisation vector for aes, password hashing salt
+    """
+    # Generate hash from password
+    salt = os.urandom(16)
+    hashed_pass = pbkdf2_hmac('sha256', password.encode(), salt, 1000000)
+
+    # Random initialisation vector
+    iv = ''.join([chr(random.randint(33, 127)) for _ in range(16)])
```

Comment:
I suppose that it doesn't really matter, but I guess this is RNG source is not as good as `os.urandom`

Reply:
Switched this to os.urandom - there was no good reason not to.